persistence/state.go

```diff
@@ -0,0 +1,187 @@
+	types.ActorType_App:  types.ApplicationActor,
+	types.ActorType_Val:  types.ValidatorActor,
+	types.ActorType_Fish: types.FishermanActor,
+	types.ActorType_Node: types.ServiceNodeActor,
+}
+
+func newMerkleTrees() (map[MerkleTree]*smt.SparseMerkleTree, error) {
+	// We need a separate Merkle tree for each type of actor or storage
+	trees := make(map[MerkleTree]*smt.SparseMerkleTree, int(lastMerkleTree))
+
+	for treeType := MerkleTree(0); treeType < lastMerkleTree; treeType++ {
+		// TODO_IN_THIS_COMMIT: Rather than using `NewSimpleMap`, use a disk based key-value store
+		nodeStore := smt.NewSimpleMap()
+		valueStore := smt.NewSimpleMap()
+
+		trees[treeType] = smt.NewSparseMerkleTree(nodeStore, valueStore, sha256.New())
+	}
+	return trees, nil
+}
+
+func loadMerkleTrees(map[MerkleTree]*smt.SparseMerkleTree, error) {
+	log.Fatalf("loadMerkleTrees not implemented yet")
+}
+
+func (p *PostgresContext) updateStateHash() error {
+	// Update all the merkle trees
+	for treeType := MerkleTree(0); treeType < lastMerkleTree; treeType++ {
+		switch treeType {
+		case appMerkleTree:
+			fallthrough
+		case valMerkleTree:
+			fallthrough
+		case fishMerkleTree:
+			fallthrough
+		case serviceNodeMerkleTree:
+			actorType, ok := merkleTreeToActorTypeName[treeType]
+			if !ok {
+				return fmt.Errorf("no actor type found for merkle tree: %v\n", treeType)
+			}
+			actors, err := p.getActorsUpdatedAtHeight(actorType, p.Height)
+			if err != nil {
+				return err
+			}
+			if err != p.updateActorsTree(actorType, actors) {
+				return err
+			}
+		case accountMerkleTree:
+			fallthrough
+		case poolMerkleTree:
+			fallthrough
+		case blocksMerkleTree:
+			fallthrough
+		case paramsMerkleTree:
+			fallthrough
+		case flagsMerkleTree:
+			// log.Println("TODO: merkle tree not implemented", treeType)
+		default:
+			log.Fatalln("Not handled yet in state commitment update", treeType)
+		}
+	}
+
+	// Get the root of each Merkle Tree
+	roots := make([][]byte, 0)
+	for treeType := MerkleTree(0); treeType < lastMerkleTree; treeType++ {
+		roots = append(roots, p.merkleTrees[treeType].Root())
+	}
+
+	// DISCUSS(drewsky): In #152, we discussed the ordering of the roots
+	// 	Strict Ordering: sha3(app_tree_root + fish_tree_root + service_node_tree_root + validator_tree_root)
+	// 	Value Ordering sha3(app_tree_root <= + fish_tree_root <= + service_node_tree_root <= + validator_tree_root)
+	// If we don't do the lexographic ordering below, then it follows the string ordering of
+	// the merkle trees declared above. I have a feeling you're not a fan of this solution, but curious
+	// to hear your thoughts.
+
+	// Get the state hash
+	rootsConcat := bytes.Join(roots, []byte{})
+	stateHash := sha256.Sum256(rootsConcat)
+
+	p.currentStateHash = stateHash[:]
+	return nil
+}
+
+func (p PostgresContext) updateActorsTree(actorType types.ActorType, actors []*types.Actor) error {
```

Comment:
Since this is a partial implementation - I won't dig into it - but wanted to **emphasize** that the `world state` is a lot more than just the `actors`.